### PR TITLE
add load_forecasts_zoltar() function

### PR DESCRIPTION
**Pull request by matthewcornell**
closes [create a zoltar option in connect_hub() #10]

### CODE REVIEW

File: R/load_forecasts_zoltar.R
Comment (bsweger) [inline R/load_forecasts_zoltar.R:1]: What do you think about giving this function a name that more closely matches the existing `hubData` interface (_e.g._, `connect_hub_zoltar`, `connect_zoltar`, `connect_model_output_zoltar`)?

**Review by bsweger — COMMENTED** [review]
Very excited to see `hubData` expanded to include a Zoltar connection!

I had one inline comment about the function name. Also noting that I was able to use the function successfully as documented, but got a 503 when connecting to a large hub and grabbing everything.

```
> df <- load_forecasts_zoltar("ECDC European COVID-19 Forecast Hub")
[snip]
get_resource(): GET: https://zoltardata.com/api/job/616053/
QUEUED
get_resource(): GET: https://zoltardata.com/api/job/616053/
QUEUED
get_resource(): GET: https://zoltardata.com/api/job/616053/
SUCCESS
get_resource(): GET: https://zoltardata.com/api/job/616053/data/
Error in get_resource(zoltar_connection, data_url, col_types) : 
  Service Unavailable (HTTP 503).
```

**Review by nickreich — CHANGES_REQUESTED** [review]
I have not reviewed the code in detail but @matthewcornell asked for input on documentation. I reviewed the [netlify documentation preview of the load_forecasts_zoltar() function](https://6669a93550c0d7349aaaf9b7--hubdata-pr-preview.netlify.app/reference/load_forecasts_zoltar) and have a few minor suggestions/comments.

1. The details section of the page looks like it is supposed to render as a list but instead it just reads as one paragraph. This should be edited so it renders as a list
2. Add a bit more info (2-3 sentences) about what Zoltar is in the documentation. A hubverse user might have no idea what Zoltar is, so we need to tell them what it is, where it lives (adding an actual link, not just text). I'm thinking...
3. In addition to a general introduction, include a few details about how zoltar concepts map to hubverse concepts, like how "units" and "targets" are like "task-id" variables. and "timezero" is a round id? or something like that.
4. add a link/reference to zoltr, so people could understand how to interact with and extract data from that tool.
5. I think it should be clearer in the documentation that **you need a Zoltar account for this function to work for you**. Right now, this is stated kind of implicitly in the details when talking about environment variables.

In terms of adding an example of this code to the vignette, I this this is not necessary for this PR because it adds complexity (e.g. storing secrets) and won't be reproducible for users without Zoltar accounts. But I will say that I large advantage of this function is how it will make the covid-19 forecast hub data directly available in hubverse format, and maybe giving a demo of that could be useful somewhere at some point. Maybe worth filing an issue here, or in the main hubverse repo, at some point about that.

File: R/load_forecasts_zoltar.R
Comment (matthewcornell) [reply R/load_forecasts_zoltar.R:1]: Re: 503, yes, it will time out if you ask for a "large" number of forecasts. That particular project has ~1/2M predictions, each of which has ~20 data rows, for ~10M rows. I believe the workers that handle queries are limited to a 7 minute runtime before they're shut down.

Re: the function name, I'll leave it to others to decide (@nickreich , @lshandross , @annakrystalli  ?)

Comment (lshandross) [reply R/load_forecasts_zoltar.R:1]: I'm happy with either of the first two names (the last one feels too wordy)

**Comment by matthewcornell** [comment]
@nickreich and @bsweger , @lshandross and I have updated the code to incorporate your helpful comments. We renamed the function and updated its docs. Please take a look - the [netlify preview](https://6669f8753bf1502267f2b365--hubdata-pr-preview.netlify.app/reference/collect_zoltar_forecasts) shows the final function documentation.

File: R/collect_zoltar_forecasts.R [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+                                         stringr::str_detect(.data$value, "[^\\d|\\.]", negate = TRUE),
+                                         TRUE),
+                  numeric_sample = ifelse(!is.na(sample),
+                                          stringr::str_detect(sample, "[^\\d|\\.]", negate = TRUE),
+                                          TRUE)) |>
+    dplyr::filter(.data$numeric_value, .data$numeric_sample, !class %in% c("named", "mode")) |>
+    dplyr::mutate(value = as.numeric(.data$value)) |>
+    dplyr::mutate(hub_target = stringr::str_squish(stringr::str_remove(.data$target, paste0(.data$numeric_horizon)))) |>
+    dplyr::group_split(class) |>
+    purrr::map_dfr(.f = function(split_outputs) {
```
Comment (annakrystalli) [inline R/collect_zoltar_forecasts.R:104]: Note that `map_dfr()` has [been superceded in `purrr` in favour of `map() |> purrr::list_rbind()`](https://purrr.tidyverse.org/reference/map_dfr.html)

File: R/collect_zoltar_forecasts.R [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+#'
+#' `collect_zoltar_forecasts` retrieves data from a \href{https://zoltardata.com/}{zoltardata.com} project and
+#' transforms it from Zoltar's native download format into a hubverse one. Zoltar (documentation
+#' \href{https://docs.zoltardata.com/}{here}) is a pre-hubverse research project that implements a repository of model
+#' forecast results, including tools to administer, query, and visualize uploaded data, along with R and Python APIs to
+#' access data programmatically (\href{https://github.com/reichlab/zoltr/}{zoltr} and
+#' \href{https://github.com/reichlab/zoltpy/}{zoltpy}, respectively.) (This hubData function is itself implemented using
+#' the zoltr package.)
+#'
+#' @param project_name a string naming the Zoltar project to load forecasts from. assumes the host is zoltardata.com
```
Comment (annakrystalli) [inline R/collect_zoltar_forecasts.R:13]: General comment on the documentation. It should have proper sentence format (i.e. capitals after periods, not all lower case)

File: R/collect_zoltar_forecasts.R [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+#'   which queries all targets in the project
+#' @param types a character vector that specifies the forecast types to query. Choices are "bin", "point", "sample",
+#'   "quantile", "mean", and "median". defaults to NULL, which queries all types in the project. note: while Zoltar
+#'   supports "named" and "mode" forecasts, this function ignores them
+#' @param as_of a datetime string that specifies the forecast version. The as_of field format must be a datetime as
+#'   parsed by the \href{https://dateutil.readthedocs.io/}{dateutil python library}, which accepts a variety of styles.
+#'   You can find examples \href{https://dateutil.readthedocs.io/en/stable/examples.html#parse-examples}{here}.
+#'   importantly, the datetime must include timezone information for disambiguation, without which the query will fail.
+#'   also, the datatime parsing function used below is extremely lenient when it comes to formatting, so please exercise
+#'   caution. defaults to NULL to load the latest version
```
Comment (annakrystalli) [inline R/collect_zoltar_forecasts.R:30]: Why not make things easier by requiring that users provide character strings of dates in ISO format only? You can then check that its valid and then safely convert to a format zoltar will accept? Makes it simpler to document too. 

If not, given you are using `strftime(as_of, format = "%Y-%m-%d %H:%M:%S", tz = "UTC", usetz = TRUE)` to check whether the input is valid, it's better to point to that as the expectation in the docs rather to a python package as that's what's actually going to trigger the failure.

File: R/collect_zoltar_forecasts.R [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+#'                         targets = c("pct next week", "cases next week"), types = c("point"),
+#'                         as_of = NULL, point_output_type = "mean")
+#' }
+#'
+#' @importFrom rlang .data
+#' @importFrom zoltr do_zoltar_query
+collect_zoltar_forecasts <- function(project_name, models = NULL, timezeros = NULL, units = NULL, targets = NULL,
+                                     types = NULL, as_of = NULL, point_output_type = "median") {
+  zoltar_connection <- zoltr::new_connection()
+  zoltr::zoltar_authenticate(zoltar_connection, Sys.getenv("Z_USERNAME"), Sys.getenv("Z_PASSWORD"))
```
Comment (annakrystalli) [inline R/collect_zoltar_forecasts.R:70]: Is there any useful metadata made available through a zoltar connection or computational efficiency (if we wanted to make multliple subsequent but different queries to a project) to be gained from having the above two steps separated into a `connect_zoltar()` function?

File: R/collect_zoltar_forecasts.R [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+format_to_hub_model_output <- function(forecasts, zoltar_targets_df, point_output_type) {
+  hub_model_outputs <- forecasts |>
+    dplyr::left_join(dplyr::select(zoltar_targets_df, .data$name, .data$numeric_horizon), by = c("target" = "name")) |>
+    dplyr::mutate(numeric_value = ifelse(!is.na(.data$value),
+                                         stringr::str_detect(.data$value, "[^\\d|\\.]", negate = TRUE),
+                                         TRUE),
+                  numeric_sample = ifelse(!is.na(sample),
+                                          stringr::str_detect(sample, "[^\\d|\\.]", negate = TRUE),
+                                          TRUE)) |>
+    dplyr::filter(.data$numeric_value, .data$numeric_sample, !class %in% c("named", "mode")) |>
```
Comment (annakrystalli) [inline R/collect_zoltar_forecasts.R:100]: Although the tests pass, am noticng a bunch of warning in the output re deprecation of `.data$` in tidyselect since 1.2.0, e.g. 
https://github.com/Infectious-Disease-Modeling-Hubs/hubData/actions/runs/9488629904/job/26147999504?pr=36#step:6:161

This will likely affect other code in this package and other packages as well but it's probably good to understand what the up to date way of using tidy select is and start applying it throughout our packages.  Found this[ blogpost which seems useful](https://www.tidyverse.org/blog/2022/10/tidyselect-1-2-0/). Looks like we need to use different approaches in selecting situations vs data masking (e.g. mutating) situations.

Comment (annakrystalli) [reply R/collect_zoltar_forecasts.R:100]: Also, for simple stuff like selecting a couple of columns, it's easier to use base R.

File: R/load_forecasts_zoltar.R
Comment (annakrystalli) [reply R/load_forecasts_zoltar.R:1]: I would actually advocate for `collect_zoltar()` for more consistency with `collect_hub()` and more conciseness (if that's even a word !)

**Comment by annakrystalli** [comment]
Final points:
- [ ] The version needs to be bumped in the `DESCRIPTION` file
- [ ] An item introducing the new feature needs to be added in `NEWS.md`

**Review by annakrystalli — CHANGES_REQUESTED** [review]

**Comment by matthewcornell** [comment]
@nickreich said:
> It probably is worth updating CRAN. 0.5.1 on CRAN vs. 1.0.1 on GitHub.

**Review by nickreich — COMMENTED** [review]
I second @annakrystalli 's stylistic requests for the documentation (sentence capitalization, function re-naming, etc...), but my documentation-content related concerns from my first review have been addressed very nicely with the additional descriptive/background/informational text in the documentation. thank you!

File: R/load_forecasts_zoltar.R
Comment (bsweger) [reply R/load_forecasts_zoltar.R:1]: > Re: 503, yes, it will time out if you ask for a "large" number of forecasts. That particular project has ~1/2M predictions, each of which has ~20 data rows, for ~10M rows. I believe the workers that handle queries are limited to a 7 minute runtime before they're shut down. You could try a more constrained query, such as models, timezeros, or units.
> 

Got it! Then I would vote for catching 5xx HTTP errors and returning a user-friendly message about trying the connection again with more specific filters.

Comment (matthewcornell) [reply R/load_forecasts_zoltar.R:1]: > I would actually advocate for `collect_zoltar()` for more consistency with `collect_hub()` and more conciseness (if that's even a word !)

Thanks for the suggestion, Anna. We included "forecasts" in the function name because we wanted to differentiate between querying zoltar forecasts (which this function does) and querying truth (possibly in a separate function in the future).

Comment (annakrystalli) [reply R/load_forecasts_zoltar.R:1]: I see. In `collect_hub()` we are also only accessing model output data though so it still fits the design pattern. When we decide if and how target data will be accessed we can try and mirror the approach?

**Comment by matthewcornell** [comment]
Thanks for the comments, Anna! Here's what we decided:
- re: the ".data$" warnings: we changed our two dplyr::select() calls to use base R, but we still have warnings from other dplyr calls. We aren't sure how to proceed in getting rid of those remaining warnings.
- re: ISO format: Thanks for the suggestion. We decided to keep the current code rather than dive deep into new formatting, but we included information about the actual R function used, as you suggested.
- re: Is there any useful metadata: Good idea. I'll make an issue for it.
- re: showing actual query results: ""

File: R/collect_zoltar_forecasts.R (hunk 5 above)
Comment (annakrystalli) [reply R/collect_zoltar_forecasts.R]: I think this is a wider `dplyr` style guide we'll need to sit out and work out all together and start applying across our packages as the new way seems to be less intuitive initially and needs working through.

**Comment by annakrystalli** [comment]
Thanks for the updates @matthewcornell and @lshandross !
Most of my comments have been addressed.

I still feel `collect_zoltar()` is best in terms of function name (see here: https://github.com/Infectious-Disease-Modeling-Hubs/hubData/pull/36#discussion_r1638291425)

I've also made a suggested change to set a minimum `zoltr` version in `DESCRIPTION` given earlier versions don't export the required function.

**Review by annakrystalli — APPROVED** [review]

**Review by bsweger — APPROVED** [review]
This comment got lost in a thread about function names, but if we know in advance that this feature won't work when querying large volumes of data from Zoltar, it would be more polished if we caught that HTTP 503 and returned a more user-friendly message advising people to add additional filters/constraints to their request.

That could be a follow-up, however. Great to have this foundational layer in!

**Review by nickreich — APPROVED** [review]
Great work on this! Excited to have this functionality in the package.
